pkg/haprovider/haprovider.go

```diff
@@ -230,13 +230,34 @@ func (r *HAProvider) updateClusterControlPlaneEndpoint(cluster *clusterv1.Cluste
 	return errors.New(service.Name + " service external ip is not ready")
 }
 
-func (r *HAProvider) findEndpointInMachine(ip string, machine *clusterv1.Machine) bool {
+func (r *HAProvider) syncEndpointMachineIP(address corev1.EndpointAddress, machine *clusterv1.Machine) corev1.EndpointAddress {
+	// check if ip is still in the machine's address
+	found := false
 	for _, machineAddress := range machine.Status.Addresses {
-		if net.ParseIP(machineAddress.Address) != nil && ip == machineAddress.Address {
-			return true
+		if machineAddress.Type != clusterv1.MachineExternalIP {
+			continue
+		}
+		if net.ParseIP(machineAddress.Address) != nil && address.IP == machineAddress.Address {
+			found = true
 		}
 	}
-	return false
+
+	if !found {
+		// machine address is not sync with the ip address in endpoints object
```

Comment:
Just realized CPI is only installed after kubelet is ready. Wonder does it require the apiserver to be ready.
Otherwise CPI can't be running, then external IP will be empty. Maybe we can only rely on the internal IP

Reply:
CPI is setting address for `node` object, looks like it is `capv` which is setting address for `machine`.

And the machine only has `externalIP` type address in vsphere env

Reply:
Good